simfection/interaction_engine_cpp.py

```diff
@@ -105,6 +106,98 @@ def interact_all(self):
                 )
             )
 
+class InteractionEngineCpp2(InteractionEngineCpp):
+    def _prep_for_cpp(self) -> [dict,dict,dict]:
+        '''Preps data to pass off to cython functions.
+
+        Returns:
+            population: dict
+            connections: dict
+            pathogen: dict
+        '''
+        # TODO: Make sure to change __init__ to encode the keys in
+        #       the pathogen dictionary.
+        population = {
+            'agent': self.population['agent'].tolist(),
+            'state': (
+                self.population['state']
+                .apply(lambda state: state.encode())
+                .tolist(),
+            ),
+            'infected_by': self.population['infected_by'].tolist(),
+            'days_infected': self.population['days_infected'].tolist(),
+            'immunity': self.population['immunity'].tolist(),
+            # Working version
+            # pop_immunity = self.population['immunity'].tolist()
+            # NOTE: I specifically tested, and pop_immunity == popuation('immunity'.get)
+            # This also fails, though if I try to load pop_immunity from the original engine
+        }
+
+
+
+        connections = {
+            'agent': self.connections['agent'].tolist(),
+            'connections': self.connections['connections'].tolist(),
+            'num_connections': self.connections['num_connections'].tolist(),
+            'max_connections': bytearray(self.connections['max_connections'].tolist()),
+        }
+
+        pathogen = {k.encode():v for k,v in self.pathogen.items()}
+
+        return population, connections, pathogen
+
+    @staticmethod
+    def _set_interactions(population,connections,pathogen):
+        '''Returns the PyInteractions instance with population and
+        connections already set.'''
+        interactions = PyInteractions()
+        # TODO: Update method to accept dictionary
+        interactions.set_population(
+            population.get('agent'),
```

Comment:
Would probably recommend using the same method for accessing dict members that you use in `interactions.set_connections`, or vice versa. Might not matter, but there could be differences in the implementation there. PyInteractions should automatically convert list[int] to vector[int] and back to list[int] when passing between Python <-> Cpp. I think I might actually prefer to use dict['key'] to get the values, since dict.get() returns none (since no default is here) if the key is not found, and there is no None type equivalent in Cpp.

Reply:
It's actually consider "best practices" to avoid the `dict['key']` pattern apparently, precisely because it returns something if the key is not found. 

We have the ability to change what is returned if the key cannot be found using the `default` keyword argument `some_dict.get('key',default=-1)`.  (I think I've read it's common to use some placeholder like -1 for `Nonetype` in languages like C.) 

Either way, you are correct, we should be consistent throughout. I just recently read about the `get` method for dictionaries, so I haven't fully implemented it yet.

Reply:
You should use the method that has the implementation with the behavior you desire. Sometimes it is better to have a script fail or return an exception as opposed to return None type. If you want to return None type, or another default value, then use dict.get(). It might be easier to see when it fails with dict['key'] because you get a key value error exception, and since this isn't something that accepts user input, it might be better for it to fail explicitly rather than silently if someone is trying to program in the wrong key name.

Reply:
@gvermillion I would recommend using dict[‘key’] method here, since we’re not handling user input and it should fail hard if programmers put in the wrong key name. It will be easier to debug a key error than some downstream PyInteractions error (which tends to be more obscure).